### Load data

In [3]:
import pandas as pd
import seaborn as sns
import numpy as np

In [5]:
df = pd.read_csv('all-data.csv', delimiter=';', encoding= 'unicode_escape')
df.shape

(4292, 26)

In [6]:
df.head(5)

Patient ID Age recode with <1 year olds Race/ethnicity     Sex  \
0       18799                  15-19 years          White    Male   
1       70613                  10-14 years          White  Female   
2       81498                  15-19 years          White    Male   
3       83124                  15-19 years          White    Male   
4      106926                  10-14 years          White  Female   

   Year of diagnosis Site recode ICD-O-3/WHO 2008  \
0               1980             Bones and Joints   
1               1977             Bones and Joints   
2               1980             Bones and Joints   
3               1979             Bones and Joints   
4               1979             Bones and Joints   

  SEER historic stage A (1973-2015)  \
0                          Regional   
1                           Distant   
2                         Localized   
3                         Localized   
4                         Localized   

               Reason no cancer-directed surgery  \
0  Recommended but not performed, unknown reason   
1  Recommended but not performed, unknown reason   
2  Recommended but not performed, unknown reason   
3  Recommended but not performed, unknown reason   
4                              Surgery performed   

  2-Digit NS EOD part 1 (1973-1982) 2-Digit SS EOD part 1 (1973-1982)  ...  \
0                          Blank(s)                                 7  ...   
1                          Blank(s)                                 &  ...   
2                          Blank(s)                                 4  ...   
3                          Blank(s)                                 4  ...   
4                          Blank(s)                                 4  ...   

  Regional nodes examined (1988+) Regional nodes positive (1988+)  \
0                        Blank(s)                        Blank(s)   
1                        Blank(s)                        Blank(s)   
2                        Blank(s)                        Blank(s)   
3                        Blank(s)                        Blank(s)   
4                        Blank(s)                        Blank(s)   

  Survival months       COD to site recode  \
0              28         Bones and Joints   
1              17         Bones and Joints   
2             323                    Alive   
3             462  Kidney and Renal Pelvis   
4             472                    Alive   

  SEER cause-specific death classification             Sequence number  \
0    Dead (attributable to this cancer dx)            One primary only   
1    Dead (attributable to this cancer dx)            One primary only   
2             Alive or dead of other cause            One primary only   
3    Dead (attributable to this cancer dx)            One primary only   
4             Alive or dead of other cause  1st of 2 or more primaries   

  Total number of in situ/malignant tumors for patient  \
0                                                  1     
1                                                  1     
2                                                  1     
3                                                  1     
4                                                  2     

  Total number of benign/borderline tumors for patient  \
0                                                  0     
1                                                  0     
2                                                  0     
3                                                  0     
4                                                  0     

                  Type of Reporting Source  \
0  Hospital inpatient/outpatient or clinic   
1  Hospital inpatient/outpatient or clinic   
2  Hospital inpatient/outpatient or clinic   
3  Hospital inpatient/outpatient or clinic   
4  Hospital inpatient/outpatient or clinic   

  Median household income inflation adj to 2019  
0        Unknown/missing/no match/Not 1990-2018  
1        Unknown/missing/no match/Not 19

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4292 entries, 0 to 4291
Data columns (total 26 columns):
 #   Column                                                Non-Null Count  Dtype 
---  ------                                                --------------  ----- 
 0   Patient ID                                            4292 non-null   int64 
 1   Age recode with <1 year olds                          4292 non-null   object
 2   Race/ethnicity                                        4292 non-null   object
 3   Sex                                                   4292 non-null   object
 4   Year of diagnosis                                     4292 non-null   int64 
 5   Site recode ICD-O-3/WHO 2008                          4292 non-null   object
 6   SEER historic stage A (1973-2015)                     4292 non-null   object
 7   Reason no cancer-directed surgery                     4292 non-null   object
 8   2-Digit NS EOD part 1 (1973-1982)                     4292 non-null 

### Drop duplicates and get the 2581 rows in data

In [8]:
temp = df.drop_duplicates()
temp.shape

(2581, 26)

In [9]:
df = temp


In [10]:
temp.reset_index(drop=True)
df.reset_index(drop=True)

Patient ID Age recode with <1 year olds        Race/ethnicity     Sex  \
0          18799                  15-19 years                 White    Male   
1          70613                  10-14 years                 White  Female   
2          81498                  15-19 years                 White    Male   
3          83124                  15-19 years                 White    Male   
4         106926                  10-14 years                 White  Female   
...          ...                          ...                   ...     ...   
2576    75756574                  01-04 years                 White  Female   
2577    75803341                  05-09 years                 White  Female   
2578    75825234                  05-09 years                 White    Male   
2579    75836356                  05-09 years                 White  Female   
2580    75872838                  15-19 years  Asian Indian (2010+)  Female   

      Year of diagnosis Site recode ICD-O-3/WHO 2008  \
0                  1980             Bones and Joints   
1                  1977             Bones and Joints   
2                  1980             Bones and Joints   
3                  1979             Bones and Joints   
4                  1979             Bones and Joints   
...                 ...                          ...   
2576               2016             Bones and Joints   
2577               2017             Bones and Joints   
2578               2017  Soft Tissue including Heart   
2579               2018             Bones and Joints   
2580               2018  Soft Tissue including Heart   

     SEER historic stage A (1973-2015)  \
0                             Regional   
1                              Distant   
2                            Localized   
3                            Localized   
4                            Localized   
...                                ...   
2576                          Blank(s)   
2577                          Blank(s)   
2578                          Blank(s)   
2579                          Blank(s)   
2580                          Blank(s)   

                  Reason no cancer-directed surgery  \
0     Recommended but not performed, unknown reason   
1     Recommended but not performed, unknown reason   
2     Recommended but not performed, unknown reason   
3     Recommended but not performed, unknown reason   
4                                 Surgery performed   
...                                             ...   
2576                                Not recommended   
2577                              Surgery performed   
2578                                Not recommended   
2579                              Surgery performed   
2580                                Not recommended   

     2-Digit NS EOD part 1 (1973-1982) 2-Digit SS EOD part 1 (1973-1982)  ...  \
0                             Blank(s)                                 7  ...   
1                             Blank(s)                                 &  ...   
2                             Blank(s)                                 4  ...   
3                             Blank(s)                                 4  ...   
4                             Blank(s)                                 4  ...   
...                                ...                               ...  ...   
2576                          Blank(s)                          Blank(s)  ...   
2577                          Blank(s)                          Blank(s)  ...   
2578                          Blank(s)                          Blank(s)  ...   
2579                          Blank(s)                          Blank(s)  ...   
2580                          Blank(s)                          Blank(s)  ...   

     Regional nodes examined (1988+) Regional nodes positive (1988+)  \
0                           Blank(s)                        Blank(s)   
1                           Blank(s)                        Blank(s)   
2                           B

### Save data with no duplicates

In [12]:
df.to_csv('no-duplicates-data.csv')

### Column tummor size

#### data in columns: EOD 4 size, CS tumoe size, Tumor size Summary shows size of tummor of patients
#### i need just one column to have data about tummor size

In [43]:
df.isna().sum()

Patient ID                                                 0
Age recode with <1 year olds                               0
Race/ethnicity                                            13
Sex                                                        0
Year of diagnosis                                          0
Site recode ICD-O-3/WHO 2008                               0
SEER historic stage A (1973-2015)                        273
Reason no cancer-directed surgery                          0
2-Digit NS EOD part 1 (1973-1982)                       2557
2-Digit SS EOD part 1 (1973-1982)                       2433
EOD 4 - size (1983-1987)                                2459
EOD 4 - extent (1983-1987)                              2459
EOD 4 - nodes (1983-1987)                               2459
EOD 10 - size (1988-2003)                               1769
CS tumor size (2004-2015)                               1380
Tumor Size Summary (2016+)                              2318
Regional nodes examined 

In [24]:
df = df.sort_values(by=["Year of diagnosis"])


In [17]:
df['Tumor size'] = df['EOD 4 - size (1983-1987)']
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2581 entries, 0 to 4291
Data columns (total 27 columns):
 #   Column                                                Non-Null Count  Dtype 
---  ------                                                --------------  ----- 
 0   Patient ID                                            2581 non-null   int64 
 1   Age recode with <1 year olds                          2581 non-null   object
 2   Race/ethnicity                                        2581 non-null   object
 3   Sex                                                   2581 non-null   object
 4   Year of diagnosis                                     2581 non-null   int64 
 5   Site recode ICD-O-3/WHO 2008                          2581 non-null   object
 6   SEER historic stage A (1973-2015)                     2581 non-null   object
 7   Reason no cancer-directed surgery                     2581 non-null   object
 8   2-Digit NS EOD part 1 (1973-1982)                     2581 non-null 

In [25]:
df = df.reset_index(drop=True)
df.to_csv('sorted-data.csv')

In [19]:
df = df.replace('Unknown', np.nan)
df = df.replace('Blank(s)', np.nan)
df = df.replace('Unknown/missing/no match/Not 1990-2018', np.nan)
df

Patient ID Age recode with <1 year olds        Race/ethnicity     Sex  \
0          18799                  15-19 years                 White    Male   
1          70613                  10-14 years                 White  Female   
2          81498                  15-19 years                 White    Male   
3          83124                  15-19 years                 White    Male   
4         106926                  10-14 years                 White  Female   
...          ...                          ...                   ...     ...   
4287    75756574                  01-04 years                 White  Female   
4288    75803341                  05-09 years                 White  Female   
4289    75825234                  05-09 years                 White    Male   
4290    75836356                  05-09 years                 White  Female   
4291    75872838                  15-19 years  Asian Indian (2010+)  Female   

      Year of diagnosis Site recode ICD-O-3/WHO 2008  \
0                  1980             Bones and Joints   
1                  1977             Bones and Joints   
2                  1980             Bones and Joints   
3                  1979             Bones and Joints   
4                  1979             Bones and Joints   
...                 ...                          ...   
4287               2016             Bones and Joints   
4288               2017             Bones and Joints   
4289               2017  Soft Tissue including Heart   
4290               2018             Bones and Joints   
4291               2018  Soft Tissue including Heart   

     SEER historic stage A (1973-2015)  \
0                             Regional   
1                              Distant   
2                            Localized   
3                            Localized   
4                            Localized   
...                                ...   
4287                               NaN   
4288                               NaN   
4289                               NaN   
4290                               NaN   
4291                               NaN   

                  Reason no cancer-directed surgery  \
0     Recommended but not performed, unknown reason   
1     Recommended but not performed, unknown reason   
2     Recommended but not performed, unknown reason   
3     Recommended but not performed, unknown reason   
4                                 Surgery performed   
...                                             ...   
4287                                Not recommended   
4288                              Surgery performed   
4289                                Not recommended   
4290                              Surgery performed   
4291                                Not recommended   

     2-Digit NS EOD part 1 (1973-1982) 2-Digit SS EOD part 1 (1973-1982)  ...  \
0                                  NaN                                 7  ...   
1                                  NaN                                 &  ...   
2                                  NaN                                 4  ...   
3                                  NaN                                 4  ...   
4                                  NaN                                 4  ...   
...                                ...                               ...  ...   
4287                               NaN                               NaN  ...   
4288                               NaN                               NaN  ...   
4289                               NaN                               NaN  ...   
4290                               NaN                               NaN  ...   
4291                               NaN                               NaN  ...   

     Regional nodes positive (1988+) Survival months       COD to site recode  \
0                                NaN              28         Bones and Joints   
1                                NaN              17         Bones and Joints   
2 

In [38]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2581 entries, 0 to 2580
Data columns (total 27 columns):
 #   Column                                                Non-Null Count  Dtype  
---  ------                                                --------------  -----  
 0   Patient ID                                            2581 non-null   int64  
 1   Age recode with <1 year olds                          2581 non-null   object 
 2   Race/ethnicity                                        2568 non-null   object 
 3   Sex                                                   2581 non-null   object 
 4   Year of diagnosis                                     2581 non-null   int64  
 5   Site recode ICD-O-3/WHO 2008                          2581 non-null   object 
 6   SEER historic stage A (1973-2015)                     2308 non-null   object 
 7   Reason no cancer-directed surgery                     2581 non-null   object 
 8   2-Digit NS EOD part 1 (1973-1982)                     24 n

### add data from column CS Tumor size in column Tumor size

In [34]:
df['EOD 4 - size (1983-1987)'] = df['EOD 4 - size (1983-1987)'].astype(float)
df['EOD 10 - size (1988-2003)'] = df['EOD 10 - size (1988-2003)'].astype(float)
df['CS tumor size (2004-2015)'] = df['CS tumor size (2004-2015)'].astype(float)
df['Tumor Size Summary (2016+)'] = df['Tumor Size Summary (2016+)'].astype(float)
df['Tumor size'] = df['Tumor size'].astype(float)

In [45]:
df.loc[(df['EOD 10 - size (1988-2003)'].notna()), 'Tumor size'] = df['EOD 10 - size (1988-2003)']


In [46]:
df.loc[(df['CS tumor size (2004-2015)'].notna()), 'Tumor size'] = df['CS tumor size (2004-2015)']
df.loc[(df['Tumor Size Summary (2016+)'].notna()), 'Tumor size'] = df['Tumor Size Summary (2016+)']

In [50]:
df.isna().sum()

Patient ID                                                 0
Age recode with <1 year olds                               0
Race/ethnicity                                            13
Sex                                                        0
Year of diagnosis                                          0
Site recode ICD-O-3/WHO 2008                               0
SEER historic stage A (1973-2015)                        273
Reason no cancer-directed surgery                          0
2-Digit NS EOD part 1 (1973-1982)                       2557
2-Digit SS EOD part 1 (1973-1982)                       2433
EOD 4 - extent (1983-1987)                              2459
EOD 4 - nodes (1983-1987)                               2459
Regional nodes examined (1988+)                          305
Regional nodes positive (1988+)                          305
Survival months                                            4
COD to site recode                                         0
SEER cause-specific deat

In [49]:
df = df.drop(['EOD 4 - size (1983-1987)','EOD 10 - size (1988-2003)','CS tumor size (2004-2015)','Tumor Size Summary (2016+)'], axis = 1)

In [51]:
df.to_excel ('excel-data.xlsx', index = False, header=True)

### Now, i have only one column for tumor size, with 183 missing values

In [90]:
df.to_csv('fixed-data.csv')

In [54]:
df['Reason no cancer-directed surgery'].value_counts()

Surgery performed                                                               1571
Not recommended                                                                  707
Recommended but not performed, unknown reason                                    222
Unknown; death certificate; or autopsy only (2003+)                               43
Recommended, unknown if performed                                                 18
Not recommended, contraindicated due to other cond; autopsy only (1973-2002)      13
Recommended but not performed, patient refused                                     7
Name: Reason no cancer-directed surgery, dtype: int64

In [55]:
df = df.replace('Surgery performed', 'Yes')
df = df.replace('Not recommended','No')
df = df.replace('Recommended but not performed, unknown reason','No')
df = df.replace('Not recommended, contraindicated due to other cond; autopsy only (1973-2002)','No')
df = df.replace('Recommended but not performed, patient refused','No')
df = df.replace('Recommended, unknown if performed','Unknown')
df = df.replace('Unknown; death certificate; or autopsy only (2003+)','Unknown')

In [58]:
df['Reason no cancer-directed surgery'].value_counts()

Yes        1571
No          949
Unknown      61
Name: Reason no cancer-directed surgery, dtype: int64

In [66]:
df = pd.read_csv('fixed-data.csv', delimiter=',', encoding= 'unicode_escape')
df = df.drop('Unnamed: 0', axis = 1)
df

Patient ID Age recode with <1 year olds        Race/ethnicity     Sex  \
0       22929427                  15-19 years                 White    Male   
1       10123011                  05-09 years                 White    Male   
2       28016877                  15-19 years                 White    Male   
3       10055225                  15-19 years                 White    Male   
4        5166438                  15-19 years                 White    Male   
...          ...                          ...                   ...     ...   
2576    46141605                  01-04 years                 White    Male   
2577    46144083                  15-19 years                 White    Male   
2578    46146654                  01-04 years                 White    Male   
2579    46215234                  10-14 years                 White    Male   
2580    75872838                  15-19 years  Asian Indian (2010+)  Female   

      Year of diagnosis Site recode ICD-O-3/WHO 2008  \
0                  1975             Bones and Joints   
1                  1975             Bones and Joints   
2                  1975             Bones and Joints   
3                  1975             Bones and Joints   
4                  1975             Bones and Joints   
...                 ...                          ...   
2576               2018             Bones and Joints   
2577               2018             Bones and Joints   
2578               2018             Bones and Joints   
2579               2018             Bones and Joints   
2580               2018  Soft Tissue including Heart   

     SEER historic stage A (1973-2015) Reason no cancer-directed surgery  \
0                             Regional                                No   
1                             Regional                                No   
2                              Distant                                No   
3                             Regional                                No   
4                             Unstaged                               Yes   
...                                ...                               ...   
2576                               NaN                               Yes   
2577                               NaN                               Yes   
2578                               NaN                               Yes   
2579                               NaN                                No   
2580                               NaN                                No   

     2-Digit NS EOD part 1 (1973-1982) 2-Digit SS EOD part 1 (1973-1982)  ...  \
0                                  NaN                                 7  ...   
1                                  NaN                                 7  ...   
2                                  NaN                                 &  ...   
3                                  NaN                                 7  ...   
4                                  NaN                        Unknown(-)  ...   
...                                ...                               ...  ...   
2576                               NaN                               NaN  ...   
2577                               NaN                               NaN  ...   
2578                               NaN                               NaN  ...   
2579                               NaN                               NaN  ...   
2580                               NaN                               NaN  ...   

      Regional nodes positive (1988+)  Survival months  COD to site recode  \
0                                 NaN             31.0    Bones and Joints   
1                                 NaN             23.0    Bones and Joints   
2                                 NaN             28.0    Bones and Joints   
3                                 NaN             11.0    Bones and Joints   
4                                 NaN             31.0    Bones and Joints   
...                           

In [67]:
df['Site recode ICD-O-3/WHO 2008'].value_counts()

Bones and Joints                                     1879
Soft Tissue including Heart                           558
Kidney and Renal Pelvis                                27
Miscellaneous                                          16
Trachea, Mediastinum and Other Respiratory Organs      13
Retroperitoneum                                        11
Lung and Bronchus                                      10
Ovary                                                   8
Nose, Nasal Cavity and Middle Ear                       7
Other Non-Epithelial Skin                               5
Small Intestine                                         5
Other Endocrine including Thymus                        4
Pleura                                                  4
Salivary Gland                                          4
Peritoneum, Omentum and Mesentery                       4
Pancreas                                                3
Vulva                                                   3
Testis        

In [68]:
df.loc[(df['Site recode ICD-O-3/WHO 2008'] != 'Bones and Joints') & (df['Site recode ICD-O-3/WHO 2008'] != 'Soft Tissue including Heart') & (df['Site recode ICD-O-3/WHO 2008'] != 'Kidney and Renal Pelvis'), 'Site recode ICD-O-3/WHO 2008'] = 'Other'

In [69]:
df['Site recode ICD-O-3/WHO 2008'].value_counts()

Bones and Joints               1879
Soft Tissue including Heart     558
Other                           117
Kidney and Renal Pelvis          27
Name: Site recode ICD-O-3/WHO 2008, dtype: int64

In [116]:
df.to_csv('fixed2-data.csv')

In [71]:
df['Total number of in situ/malignant tumors for patient'].value_counts()

1    2436
2     129
3      15
4       1
Name: Total number of in situ/malignant tumors for patient, dtype: int64

In [72]:
df['SEER historic stage A (1973-2015)'].value_counts()

Regional     824
Localized    679
Distant      648
Unstaged     157
Name: SEER historic stage A (1973-2015), dtype: int64

In [73]:
df['Sequence number'].value_counts()

One primary only              2411
1st of 2 or more primaries     133
2nd of 2 or more primaries      35
3rd of 3 or more primaries       2
Name: Sequence number, dtype: int64

#### Stage is 'Localized' if patient has only one primaru tumor

In [76]:
df.loc[(df['SEER historic stage A (1973-2015)'] == 'Unstaged') & (df['Sequence number'] == 'One primary only'), 'SEER historic stage A (1973-2015)'] ='Localized'

In [77]:
df['SEER historic stage A (1973-2015)'].value_counts()

Localized    828
Regional     824
Distant      648
Unstaged       8
Name: SEER historic stage A (1973-2015), dtype: int64

In [79]:
df[(df == 'Unstaged').any(axis=1)]

Patient ID Age recode with <1 year olds Race/ethnicity     Sex  \
28       5873337                  10-14 years          White  Female   
101     29842877                  10-14 years          White  Female   
464       637582                     00 years          White  Female   
948     60376238                  05-09 years          White  Female   
1469     6117880                  15-19 years          White    Male   
1976    36090820                  15-19 years          White    Male   
2141    75567952                  10-14 years          White  Female   
2147    44022265                  15-19 years          White    Male   

      Year of diagnosis Site recode ICD-O-3/WHO 2008  \
28                 1976             Bones and Joints   
101                1980  Soft Tissue including Heart   
464                1993                        Other   
948                2002             Bones and Joints   
1469               2007             Bones and Joints   
1976               2012  Soft Tissue including Heart   
2141               2014             Bones and Joints   
2147               2014  Soft Tissue including Heart   

     SEER historic stage A (1973-2015) Reason no cancer-directed surgery  \
28                            Unstaged                           Unknown   
101                           Unstaged                               Yes   
464                           Unstaged                               Yes   
948                           Unstaged                                No   
1469                          Unstaged                                No   
1976                          Unstaged                                No   
2141                          Unstaged                               Yes   
2147                          Unstaged                                No   

     2-Digit NS EOD part 1 (1973-1982) 2-Digit SS EOD part 1 (1973-1982)  ...  \
28                                 NaN                        Unknown(-)  ...   
101                                NaN                               NaN  ...   
464                                NaN                               NaN  ...   
948                                NaN                               NaN  ...   
1469                               NaN                               NaN  ...   
1976                               NaN                               NaN  ...   
2141                               NaN                               NaN  ...   
2147                               NaN                               NaN  ...   

      Regional nodes positive (1988+)  Survival months  \
28                                NaN            454.0   
101                               NaN            465.0   
464                              99.0              2.0   
948                              99.0            198.0   
1469                             98.0             42.0   
1976                             99.0             27.0   
2141                             99.0             54.0   
2147                             98.0             59.0   

                                COD to site recode  \
28                                          Breast   
101                                          Alive   
464   In situ, benign or unknown behavior neoplasm   
948                                          Alive   
1469                              Bones and Joints   
1976                    Acute Lymphocytic Leukemia   
2141                                         Alive   
2147                                         Alive   

      SEER cause-specific death classification             Sequence number  \
28                Alive or dead of other cause  1st of 2 or more primaries   
101               Alive or dead of other cause  1st of 2 or more primaries   
464               Alive or dead of other cause  1st of 2 or more primaries   
948               Alive or dead of other cause  1st of 2 or more primaries   
1469     Dead (attributable to this c

### There is 8 rows with 'Unstaged' stage.  All patients have 0 benign tumors
#### If patient has 2 primmary tumors stage is 'Regional' and if there is 3 primary tumors, stage is 'Distant'

In [87]:
df.loc[(df['SEER historic stage A (1973-2015)'] == 'Unstaged') & (df['Total number of in situ/malignant tumors for patient'] >= 3), 'SEER historic stage A (1973-2015)'] = 'Distant'
df.loc[(df['SEER historic stage A (1973-2015)'] == 'Unstaged') & (df['Total number of in situ/malignant tumors for patient'] == 2), 'SEER historic stage A (1973-2015)'] = 'Regional'
df.loc[(df['SEER historic stage A (1973-2015)'] == 'Unstaged') & (df['Total number of in situ/malignant tumors for patient'] == 1), 'SEER historic stage A (1973-2015)'] = 'Localized'

### There is 273 patient with NaN value for stage, I will fill missing data on the same way like 'Unstaged' value

In [147]:
df.loc[(df['SEER historic stage A (1973-2015)'].isna()) & (temp['Sequence number'] == 'One primary only'), 'SEER historic stage A (1973-2015)'] = 'Localized'

In [149]:
df.loc[(df['SEER historic stage A (1973-2015)'].isna()) & (df['Total number of in situ/malignant tumors for patient'] >= 3), 'SEER historic stage A (1973-2015)'] = 'Distant'
df.loc[(df['SEER historic stage A (1973-2015)'].isna()) & (df['Total number of in situ/malignant tumors for patient'] == 2), 'SEER historic stage A (1973-2015)'] = 'Regional'
df.loc[(df['SEER historic stage A (1973-2015)'].isna()) & (df['Total number of in situ/malignant tumors for patient'] == 1), 'SEER historic stage A (1973-2015)'] = 'Localized'

In [150]:
df['SEER historic stage A (1973-2015)'].value_counts()

Localized    1097
Regional      834
Distant       650
Name: SEER historic stage A (1973-2015), dtype: int64

### Look into Race / Entity columnt

### Make only 3 groups: White, Black and other

In [91]:
df['Race/ethnicity'].value_counts()

White                                     2275
Black                                      102
American Indian/Alaska Native               24
Other Asian (1991+)                         23
Hawaiian                                    22
Filipino                                    21
Chinese                                     17
Asian Indian (2010+)                        15
Asian Indian or Pakistani, NOS (1988+)      14
Korean (1988+)                               9
Samoan (1991+)                               7
Other                                        7
Vietnamese (1988+)                           7
Pakistani (2010+)                            5
Thai (1994+)                                 5
Tongan (1991+)                               3
Pacific Islander, NOS (1991+)                3
Japanese                                     3
Laotian (1988+)                              2
Kampuchean (1988+)                           2
Guamanian, NOS (1991+)                       2
Name: Race/et

In [92]:
df.loc[(df['Race/ethnicity'] != 'White') & (df['Race/ethnicity'] != 'Black'), 'Race/ethnicity'] = 'Other'

In [93]:
df['Race/ethnicity'].value_counts()

White    2275
Other     204
Black     102
Name: Race/ethnicity, dtype: int64

In [95]:
df['Survival months'] = df['Survival months'].astype(float)

In [99]:
df.isna().sum()

Patient ID                                                 0
Age recode with <1 year olds                               0
Race/ethnicity                                             0
Sex                                                        0
Year of diagnosis                                          0
Site recode ICD-O-3/WHO 2008                               0
SEER historic stage A (1973-2015)                        273
Reason no cancer-directed surgery                          0
2-Digit NS EOD part 1 (1973-1982)                       2557
2-Digit SS EOD part 1 (1973-1982)                       2433
EOD 4 - extent (1983-1987)                              2459
EOD 4 - nodes (1983-1987)                               2459
Regional nodes examined (1988+)                          305
Regional nodes positive (1988+)                          305
Survival months                                            4
COD to site recode                                         0
SEER cause-specific deat

### Look into couse of death
#### in COD to site recode i will make only few main groups

In [108]:
df['SEER cause-specific death classification'].value_counts()

Alive or dead of other cause             1685
Dead (attributable to this cancer dx)     878
Dead (missing/unknown COD)                 18
Name: SEER cause-specific death classification, dtype: int64

In [110]:
df['COD to site recode'].value_counts()

Alive                                                      1595
Bones and Joints                                            716
Miscellaneous Malignant Cancer                               78
Soft Tissue including Heart                                  37
Brain and Other Nervous System                               22
State DC not available or state DC available but no COD      18
In situ, benign or unknown behavior neoplasm                 17
Other Cause of Death                                         15
Diseases of Heart                                            11
Acute Myeloid Leukemia                                       10
Lung and Bronchus                                             8
Accidents and Adverse Effects                                 7
Other Infectious and Parasitic Diseases including HIV         6
Non-Hodgkin Lymphoma                                          4
Suicide and Self-Inflicted Injury                             3
Breast                                  

In [117]:
df.loc[(df['COD to site recode'] != 'Alive') & (df['COD to site recode'] != 'Bones and Joints')& (df['COD to site recode'] != 'Soft Tissue including Heart') & (df['COD to site recode']!='Brain and Other Nervous System') & (df['COD to site recode']!='Other Cause of Death'), 'COD to site recode'] = 'Dead Other Site'     

In [118]:
df['COD to site recode'].value_counts()

Alive                             1595
Bones and Joints                   716
Dead Other Site                    196
Soft Tissue including Heart         37
Brain and Other Nervous System      22
Other Cause of Death                15
Name: COD to site recode, dtype: int64

In [128]:
df['SEER cause-specific death classification'].value_counts()

TypeError: 'float' object is not subscriptable

In [173]:
df.to_csv('fixed2-data.csv', index= False)
df.to_excel ('fix-excel-data.xlsx', index = False, header=True)

## Fix missing values in tummor size
### find average for this type og stage,site and cod, and replace tumor size

In [132]:
temp = df

In [137]:
df.loc[(df['Tumor size'] == 99.0) | (df['Tumor size'] == 999.0), 'Tumor size'] = np.nan

In [152]:
df.isna().sum()

Patient ID                                                 0
Age recode with <1 year olds                               0
Race/ethnicity                                             0
Sex                                                        0
Year of diagnosis                                          0
Site recode ICD-O-3/WHO 2008                               0
SEER historic stage A (1973-2015)                          0
Reason no cancer-directed surgery                          0
2-Digit NS EOD part 1 (1973-1982)                       2557
2-Digit SS EOD part 1 (1973-1982)                       2433
EOD 4 - extent (1983-1987)                              2459
EOD 4 - nodes (1983-1987)                               2459
Regional nodes examined (1988+)                          305
Regional nodes positive (1988+)                          305
Survival months                                            4
COD to site recode                                         0
SEER cause-specific deat

In [139]:
df['Tumor size'].value_counts()

100.0    83
60.0     75
50.0     73
80.0     59
70.0     59
         ..
208.0     1
228.0     1
144.0     1
225.0     1
181.0     1
Name: Tumor size, Length: 194, dtype: int64

In [386]:
df.loc[(df['SEER historic stage A (1973-2015)'] == 'Localized') & (df['COD to site recode'] == 'Alive') & (df['Site recode ICD-O-3/WHO 2008'] == 'Kidney and Renal Pelvis'), 'Tumor size'].mean()

91.33333333333333

In [387]:
df.loc[(df['Tumor size'].isna()) & (df['SEER historic stage A (1973-2015)'] == 'Localized') & (df['COD to site recode'] == 'Alive') & (df['Site recode ICD-O-3/WHO 2008'] == 'Kidney and Renal Pelvis'), 'Tumor size'] = 91

In [179]:
df = df.replace('Unknown/missing/no match/Not 1990-2018', np.nan)

In [424]:
df.to_csv('fixed2-data.csv', index= False)
df.to_excel ('fix-excel-data.xlsx', index = False, header=True)

### Look into data, rename names of columns and remove some of them

In [423]:
df

Id    Age   Race     Sex  Diagnosed                         Site  \
0     22929427  15-19  White    Male       1975             Bones and Joints   
1     10123011  05-09  White    Male       1975             Bones and Joints   
2     28016877  15-19  White    Male       1975             Bones and Joints   
3     10055225  15-19  White    Male       1975             Bones and Joints   
4      5166438  15-19  White    Male       1975             Bones and Joints   
...        ...    ...    ...     ...        ...                          ...   
2576  46141605  01-04  White    Male       2018             Bones and Joints   
2577  46144083  15-19  White    Male       2018             Bones and Joints   
2578  46146654  01-04  White    Male       2018             Bones and Joints   
2579  46215234  10-14  White    Male       2018             Bones and Joints   
2580  75872838  15-19  Other  Female       2018  Soft Tissue including Heart   

          Stage Surgery  Extend  Months               COD  \
0      Regional      No     NaN    31.0  Bones and Joints   
1      Regional      No     NaN    23.0  Bones and Joints   
2       Distant      No     NaN    28.0  Bones and Joints   
3      Regional      No     NaN    11.0  Bones and Joints   
4     Localized     Yes     NaN    31.0  Bones and Joints   
...         ...     ...     ...     ...               ...   
2576  Localized     Yes     0.0     6.0             Alive   
2577  Localized     Yes     0.0     7.0             Alive   
2578  Localized     Yes     0.0     9.0             Alive   
2579  Localized      No     0.0     3.0             Alive   
2580  Localized      No     0.0     3.0             Alive   

                                       COD2   Sequence number  \
0     Dead (attributable to this cancer dx)  One primary only   
1     Dead (attributable to this cancer dx)  One primary only   
2     Dead (attributable to this cancer dx)  One primary only   
3     Dead (attributable to this cancer dx)  One primary only   
4     Dead (attributable to this cancer dx)  One primary only   
...                                     ...               ...   
2576           Alive or dead of other cause  One primary only   
2577           Alive or dead of other cause  One primary only   
2578           Alive or dead of other cause  One primary only   
2579           Alive or dead of other cause  One primary only   
2580           Alive or dead of other cause  One primary only   

      Malignat number                                   Report  \
0                   1  Hospital inpatient/outpatient or clinic   
1                   1  Hospital inpatient/outpatient or clinic   
2                   1  Hospital inpatient/outpatient or clinic   
3                   1  Hospital inpatient/outpatient or clinic   
4                   1  Hospital inpatient/outpatient or clinic   
...               ...                                      ...   
2576                1  Hospital inpatient/outpatient or clinic   
2577                1  Hospital inpatient/outpatient or clinic   
2578                1  Hospital inpatient/outpatient or clinic   
2579                1  Hospital inpatient/outpatient or clinic   
2580                1  Hospital inpatient/outpatient or clinic   

               Property  Tumor size  
0                   NaN        94.0  
1                   NaN        94.0  
2                   NaN       104.0  
3                   NaN        94.0  
4                   NaN       142.0  
...                 ...         ...  
2576  $50,000 - $54,999        74.0  
2577  $65,000 - $69,999        16.0  
2578           $75,000+        60.0  
2579           $75,000+       105.0  
2580  $50,000 - $54,999        78.0  

[2581 rows x 17 columns]

In [392]:
df.rename(columns={'Patient ID':'Id',
                  'Age recode with <1 year olds':'Age',
                   'Race/ethnicity':'Race',
                   'Year of diagnosis':'Diagnosed',
                   'Site recode ICD-O-3/WHO 2008':'Site',
                  'SEER historic stage A (1973-2015)':'Stage',
                   'Reason no cancer-directed surgery':'Surgery',
                   'COD to site recode':'COD',
                  'Survival months':'Months',
                   'Sequence number':'sequence_num',
                  'Total number of in situ/malignant tumors for patient':'malignant_num',
                   'Median household income inflation adj to 2019':'property',
                   'Type of Reporting Source':'Report',
                  }, inplace = True)

In [394]:
df.rename(columns={'sequence_num':'Sequence number',
                  'malignant_num':'Malignat number',
                  'property':'Property',
                   'SEER cause-specific death classification':'COD2',
                  }, inplace = True)

In [404]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2581 entries, 0 to 2580
Data columns (total 20 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Id               2581 non-null   int64  
 1   Age              2581 non-null   object 
 2   Race             2581 non-null   object 
 3   Sex              2581 non-null   object 
 4   Diagnosed        2581 non-null   int64  
 5   Site             2581 non-null   object 
 6   Stage            2581 non-null   object 
 7   Surgery          2581 non-null   object 
 8   Extend           122 non-null    float64
 9   Nodes            122 non-null    float64
 10  Extends          2276 non-null   float64
 11  Nodes            2276 non-null   float64
 12  Months           2577 non-null   float64
 13  COD              2581 non-null   object 
 14  COD2             2581 non-null   object 
 15  Sequence number  2581 non-null   object 
 16  Malignat number  2581 non-null   int64  
 17  Report        

In [399]:
df = df.drop(['2-Digit NS EOD part 1 (1973-1982)','2-Digit SS EOD part 1 (1973-1982)','Total number of benign/borderline tumors for patient'], axis = 1)

In [403]:
df.rename(columns={'EOD 4 - extent (1983-1987)':'Extend',
                  'EOD 4 - nodes (1983-1987)':'Nodes',
                  'Regional nodes examined (1988+)':'Extends',
                   'Regional nodes positive (1988+)':'Nodes',
                  }, inplace = True)

In [405]:
df['Age'] = df['Age'].apply(lambda x: x.split(' ')[0])

In [410]:
df.loc[(df['Extends'].notna()), 'Extend'] = df['Extends']


In [422]:
df = df.drop(['Node','Extends'], axis = 1)